File: neural.py
import random


class neural():
    layers: list[list[float]]
    grads: list[list[float]]
    weights: list[list[list[float]]]

    learning_rate: float

    def __init__(self, each_layer_size: list[int], *, learning_rate: float, activate_func, activate_func_d):
        self.layers = [[0] * size for size in each_layer_size]
        self.grads = [[0] * size for size in each_layer_size]
        self.weights = [[[0] * size_left for _ in range(size_right)] for size_left,
                        size_right in zip(each_layer_size[:-1], each_layer_size[1:])]
        for weight in self.weights:
            for wl in weight:
                for i in range(len(wl)):
                    wl[i] = 1 - random.random() * 2
        self.each_layer_size = each_layer_size
        self.learning_rate = learning_rate
        self.activate_func = activate_func
        self.activate_func_d = activate_func_d

    def forward(self, input_data: list[float]):
        self.layers = [[0] * size for size in self.each_layer_size]
        self.layers[0] = [*input_data]
        for layer_left, layer_right, weight in zip(self.layers, self.layers[1:], self.weights):
            for i in range(len(layer_right)):
                layer_right[i] = self.activate_func(sum(
                    w * ll for w, ll in zip(weight[i], layer_left)))

    def backward(self, ideal_output: list[float], *, loss_func_d=None):
        self.grads = [[0] * size for size in self.each_layer_size]
        self.grads[-1] = [(2 / self.each_layer_size[-1] * (real - ideal) if loss_func_d is None else loss_func_d(real, ideal)) for real,
                          ideal in zip(self.layers[-1], ideal_output)]
        for layer_left, layer_right, grad_left, grad_right, weight in zip(self.layers[::-1][1:], self.layers[::-1], self.grads[::-1][1:], self.grads[::-1], self.weights[::-1]):
            trans_weight = zip(*weight)
            for i in range(len(grad_left)):
                grad_left[i] = sum(
                    self.activate_func_d(lr) * gr * w for lr, gr, w in zip(layer_right, grad_right, next(trans_weight)))
            for i, lr, gr in zip(range(len(grad_right)), layer_right, grad_right):
                weight[i] = [value - self.learning_rate * ll *
                             self.activate_func_d(lr) * gr for value, ll in zip(weight[i], layer_left)]

    def train(self, data: list[tuple[list, list]], *, times: int, log_times: int = 1):
        gap = times if log_times <= 0 else 1 if log_times > times else times // log_times
        for i in range(times):
            for input_data, ideal_output in data:
                self.forward(input_data)
                self.backward(ideal_output)
            print(
                f"loss: {sum(v ** 2 for v in self.grads[-1]) ** 0.5}, weights: {self.weights}") if (i + 1) % gap == 0 else None


def target_func(x):
    return 1 * x**2 + 2 * x + 3


if __name__ == "__main__":
    n = neural([3, 1], learning_rate=0.01,
               activate_func=lambda v: v, activate_func_d=lambda v: 1)
    n.train([([x**2, x, 1], [target_func(x)])
            for x in [-2, -1, 0, 1, 2]], times=100, log_times=5)

File: neural.ts
declare type Iterableify<T> = { [K in keyof T]: Iterable<T[K]> }

function* zip<T extends Array<any>>(
    ...toZip: Iterableify<T>
): Generator<T> {
    // Get iterators for all of the iterables.
    const iterators = toZip.map(iterable => iterable[Symbol.iterator]())

    while (true) {
        // Advances all of the iterators.
        const results = iterators.map(iterable => iterable.next())

        // If any of the iterators are done, we should stop.
        if (results.some(({ done }) => done)) {
            break
        }

        // We can assert the yield type, since we knoe none
        // of the iterators are done.
        yield results.map(({ value }) => value) as T
    }
}

function* range(stop: number) {
    let index = 0
    while (index < stop)
        yield index++
}

declare type InitOptions = { learningRate: number, activateFunc: (v: number) => number, activateFunc_d: (v: number) => number }
declare type TrainOptions = { times: number, logTimes?: number }

class Neural {
    layers: number[][]
    grads: number[][]
    weights: number[][][]

    learningRate: number
    activateFunc: (v: number) => number
    activateFunc_d: (v: number) => number

    constructor(eachLayerSize: number[], { learningRate, activateFunc, activateFunc_d }: InitOptions) {
        this.layers = eachLayerSize.map(size => new Array<number>(size).fill(0))
        this.grads = eachLayerSize.map(size => new Array<number>(size).fill(0))
        this.weights = Array.from(zip(eachLayerSize.slice(0, -1), eachLayerSize.slice(1))).map(([leftSize, rightSize]) => new Array<number[]>(rightSize).fill(new Array<number>(leftSize).fill(0)))
        this.weights.forEach(weight => weight.forEach(wr => wr.forEach((_, i, wr) => wr[i] = 1 - Math.random() * 2)))
        this.learningRate = learningRate
        this.activateFunc = activateFunc
        this.activateFunc_d = activateFunc_d
    }

    Forward(inputData: number[]): void {
        this.layers.forEach(layer => layer.fill(0))
        Object.assign(this.layers[0], inputData)
        for (const [layerLeft, layerRight, weight] of zip(this.layers, this.layers.slice(1), this.weights)) {
            for (const i of range(layerRight.length)) {
                for (const [w, ll] of zip(weight[i], layerLeft)) { layerRight[i] += w * ll }
                layerRight[i] = this.activateFunc(layerRight[i])
            }
        }
    }

    Backward(idealOutput: number[]): void {
        this.grads.forEach(grad => grad.fill(0))
        for (const i of range(this.layers.at(-1).length)) {
            for (const [real, ideal] of zip(this.layers.at(-1), idealOutput)) { this.grads.at(-1)[i] = 2 / this.layers.at(-1).length * (real - ideal) }
        }
        this.layers.reverse(), this.grads.reverse(), this.weights.reverse()
        for (const [layerLeft, layerRight, gradLeft, gradRight, weight] of zip(this.layers.slice(1), this.layers, this.grads.slice(1), this.grads, this.weights)) {
            const transWeight = zip(...weight)
            for (const i of range(gradLeft.length)) {
                for (const [lr, gr, w] of zip(layerRight, gradRight, transWeight.next().value as number[])) { gradLeft[i] += this.activateFunc_d(lr) * gr * w }
            }
            for (const [i, lr, gr] of zip(range(gradRight.length), layerRight, gradRight)) {
                for (const [j, ll] of zip(range(weight[i].length), layerLeft)) { weight[i][j] -= this.learningRate * ll * this.activateFunc_d(lr) * gr }
            }
        }
        this.layers.reverse(), this.grads.reverse(), this.weights.reverse()
    }

    Train(data: [number[], number[]][], { times, logTimes = 1 }: TrainOptions): void {
        const gap = logTimes <= 0 ? times : logTimes > times ? 1 : Math.floor(times / logTimes)
        for (const i of range(times)) {
            for (const [inputData, idealOutput] of data) {
                this.Forward(inputData)
                this.Backward(idealOutput)
            }
            if ((i + 1) % gap == 0)
                console.log(`loss: ${this.grads.at(-1).reduce((sum, value) => sum + value ** 2, 0) ** 0.5}, weights: ${this.weights}`)
        }
    }
}

function targetFunc(x: number): number {
    return 1 * x ** 2 + 2 * x + 3
}

const n = new Neural([3, 1], { learningRate: 0.01, activateFunc: (v: number) => v, activateFunc_d: (v: number) => 1 })
n.Train([-2, -1, 0, 1, 2].map(x => [[x ** 2, x, 1], [targetFunc(x)]]), { times: 100, logTimes: 5 })